aten/src/ATen/core/ivalue.cpp

```diff
@@ -470,6 +470,69 @@ std::shared_ptr<ClassType> ivalue::Object::type() const {
   return type_.type_->expect<ClassType>();
 }
 
+IValue IValue::deepcopy() const {
+  IValue::HashAliasedIValueMap memo;
+  return deepcopy(memo);
+}
+
+IValue IValue::deepcopy(
+    IValue::HashAliasedIValueMap& memo) const {
+  if (memo.count(*this)) {
+    return memo.at(*this);
+  }
+  IValue copy;
+  switch(tag) {
+    case IValue::Tag::Tensor:
+      copy = IValue(toTensor().clone());
+      break;
+    case IValue::Tag::Tuple: {
+      std::vector<IValue> copied_tuple;
+      for (const auto& e : toTuple()->elements()) {
+        copied_tuple.push_back(e.deepcopy(memo));
+      }
+      copy = IValue(ivalue::Tuple::create(copied_tuple));
+    }
+      break;
+    case IValue::Tag::GenericList: {
+      auto list = toList();
+      auto copied_list = c10::impl::GenericList(list.elementType());
+      for (IValue v : list) {
+        copied_list.push_back(v.deepcopy(memo));
+      }
+      copy = IValue(copied_list);
+    }
+      break;
+    case IValue::Tag::GenericDict: {
+      auto dict = toGenericDict();
+      auto copied_dict = c10::impl::GenericDict(dict.keyType(), dict.valueType());
+      for (const auto& entry : dict) {
+        copied_dict.insert(entry.key().deepcopy(memo), entry.value().deepcopy(memo));
+      }
+      copy = IValue(copied_dict);
+    }
+      break;
+    case IValue::Tag::Object: {
+      copy = IValue(toObject()->deepcopy(memo));
+      break;
+    case IValue::Tag::String:
+    case IValue::Tag::None:
+    case IValue::Tag::Double:
+    case IValue::Tag::Int:
+    case IValue::Tag::Bool:
+    case IValue::Tag::Device:
+    case IValue::Tag::Uninitialized:
+      copy = *this;
```

Comment:
I think String should be copied? it can be mutated. 

Also for  scalars, I would create a new IValue because we want to prevent that any side-effect between them.
copy = IValue(n) instead of copy = *this

Reply:
I think string is constant: https://codebrowser.bddppq.com/pytorch/pytorch/aten/src/ATen/core/ivalue_inl.h.html#_ZN3c106IValueC1ENSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEE

Why do we need to create a new IValue when everything here is immutable?

Reply:
Ok, I guess it is ok in common scenario. My thinking is that deepcopy should not share anything with the original. One scenario is if someone modify the ivalue directly using some JIT custom pass/or other internal mechanism then the deep copy should not have any side-effect.

Reply:
how do people modify an immutable IValue?

Reply:
It think it is Ok. We need to make sure there is no overlap as we discussed.